bitcoin/BRWalletManager.c

```diff
@@ -1038,134 +733,188 @@ _BRWalletManagerTxDeleted (void *info,
+    //            the return of this callback.
+    bwmHandleTransactionEvent(manager,
+                              manager->wallet,
+                              transaction,
+                              (BRTransactionEvent) {
+                                  BITCOIN_TRANSACTION_DELETED
+                              });
 }
 
-static void
-_BRWalletManagerTxPublished (void *info,
-                             int error) {
-    BRWalletManager manager    = ((SubmitTransactionInfo*) info)->manager;
-    BRTransaction *transaction = ((SubmitTransactionInfo*) info)->transaction;
-    free (info);
-
-    assert  (NULL != manager->client.funcWalletEvent);
-    manager->client.funcWalletEvent (manager->client.context,
-                                     manager,
-                                     manager->wallet,
-                                     (BRWalletEvent) {
-                                         BITCOIN_WALLET_TRANSACTION_SUBMITTED,
-                                         { .submitted = { transaction, error }}
-                                     });
-}
+///
+/// Mark: SyncManager Events
+///
 
-/// MARK: - Peer Manager Callbacks
+// This callback comes from the BRSyncManager. That component has no
+// inherent threading model of its own. As such, these callbacks can
+// occur on any thread (including that of the caller that triggered
+// the event).
 
 static void
-_BRWalletManagerSaveBlocks (void *info,
-                            int replace,
-                            OwnershipKept BRMerkleBlock **blocks,
-                            size_t count) {
-    BRWalletManager manager = (BRWalletManager) info;
+_BRWalletManagerSyncEvent(void * context,
+                          BRSyncManager manager,
+                          OwnershipKept BRSyncManagerEvent event) {
+    BRWalletManager bwm = (BRWalletManager) context;
+    switch (event.type) {
+        case SYNC_MANAGER_SET_BLOCKS: {
+            fileServiceClear(bwm->fileService, fileServiceTypeBlocks);
 
-    if (replace) fileServiceClear(manager->fileService, fileServiceTypeBlocks);
-    for (size_t index = 0; index < count; index++)
-        fileServiceSave (manager->fileService, fileServiceTypeBlocks, blocks[index]);
-}
+            // !!!!!!!!!
+            // no break;
+            // !!!!!!!!!
+        }
+        case SYNC_MANAGER_ADD_BLOCKS: {
+            for (size_t index = 0; index < event.u.blocks.count; index++)
+                fileServiceSave (bwm->fileService, fileServiceTypeBlocks, event.u.blocks.blocks[index]);
+            break;
+        }
 
-static void
-_BRWalletManagerSavePeers  (void *info,
-                            int replace,
-                            OwnershipKept const BRPeer *peers,
-                            size_t count) {
-    BRWalletManager manager = (BRWalletManager) info;
+        case SYNC_MANAGER_SET_PEERS: {
+            fileServiceClear(bwm->fileService, fileServiceTypePeers);
```

Comment:
Does the `fileService` need to remain in `BWM`?  Everything the file service saves (at least currently) is related to syncing and thus maybe it is appropriately in the sync manager?

Reply:
`fileService` is used by the BRWallet callbacks as well. Personally, I think `fileService` should be wrapped in a manager (BRFileManager?) and this could potentially be used internally `BRWalletManager` and `BRSyncManager`.

I'm personally of mixed feelings on this. This thought occurred to me too but I chose to keep it where it is now given the BRWallet callbacks.